### PR TITLE
Plotting the Density Variations and vertical mass flux using the mass_budget.py code

### PR DESCRIPTION
**Pull request by mayaafields**
I think I have both of the plots correct. Let me know what you think.

### CODE REVIEW

File: python_codes/mass_budget.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+# Retrieving Data from the nc files
+fname = '/home/mafields/CylconeDiagnostics/data/jupiter2d-long-main.nc'
+data = Dataset(fname, 'r')
+
+#Defining Variables
+time = data['time']
+rho_gas = data['rho'][:,:,:,0]
+pressure = data['press'][:,:,:,0]
+mass_flux = data['v1rho'][:,:]
```
Comment (luminoctum) [inline python_codes/mass_budget.py:17]: This is the horizontally averaged mass flux, <v1rho>

File: python_codes/mass_budget.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ntime, nx1, nx2 = rho_gas.shape
+
+rho_0 = rho_gas[0,:,0]
+density_variations = rho_gas - rho_0.reshape((1,nx1,1))
+
+
+horizontally_averaged_density_var = mean(density_variations, axis =2)
+
+
+heat_rate= horizontally_averaged_density_var + mass_flux
```
Comment (luminoctum) [inline python_codes/mass_budget.py]: I think this is incorrect.
horizontally_averaged_density_var  is \<rho - rho_0\> = \<rho\> - rho_0
mass_flux is \<v1rho\>.
units are different

File: python_codes/mass_budget.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+#Plotting Density Variations
+
+pavg = mean(pressure, axis = (0,2))
+Pressure, Time = meshgrid(pavg,time)
+
+fig,ax=subplots(1,1)
+
+contour_plot_density_var = ax.contourf(Time, Pressure,horizontally_averaged_density_var)
```
Comment (luminoctum) [inline python_codes/mass_budget.py]: horizontally_average_density_var is too long. Let's name it to **rhoa_avg**
**rhoa** stands for rho - rho_0
**_avg** stands for \<\>, horizontal average

File: python_codes/mass_budget.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+title(r'$\frac{\delta \rho}{\delta t}$')
```
Comment (luminoctum) [inline python_codes/mass_budget.py:38]: You can omit the **r** in front of your string literal.

File: python_codes/mass_budget.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ax.set_xlabel(r'Time(s)',fontsize = 12)
```
Comment (luminoctum) [inline python_codes/mass_budget.py:39]: Omit **r**

File: python_codes/mass_budget.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ax.set_ylabel(r'Pressure',fontsize = 12)
```
Comment (luminoctum) [inline python_codes/mass_budget.py:40]: same

File: python_codes/mass_budget.py (hunk 3 above)
Comment (luminoctum) [inline python_codes/mass_budget.py]: simplify that line 36 to:
```
h = ax.contourf(Time, Pressure, rhop_avg)
```

File: python_codes/mass_budget.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+#Plotting the Vertical Mass Flux
+
+fig,ax= subplots(1,1)
+contour_plot_mass_flux = ax.contourf(Time, Pressure, mass_flux)
+title(r'$\frac{\delta(w \rho)}{\delta z}$')
+ax.set_xlabel(r'Time(s)',fontsize = 12)
+ax.set_ylabel(r'Pressure',fontsize = 12)
+ax.set_ylim(max(pavg), min(pavg))
+ax.set_yscale('log')
+fig.colorbar(contour_plot_mass_flux)
```
Comment (luminoctum) [inline python_codes/mass_budget.py]: Then you can use
```
fig.colorbar(h)
```

File: python_codes/mass_budget.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+fig,ax= subplots(1,1)
+contour_plot_heat_rate = ax.contourf(Time, Pressure, heat_rate)
+title(r'$-<\rho \dot{q_c}>$')
+ax.set_xlabel(r'Time(s)',fontsize = 12)
+ax.set_ylabel(r'Pressure',fontsize = 12)
+ax.set_ylim(max(pavg), min(pavg))
+ax.set_yscale('log')
+fig.colorbar(contour_plot_heat_rate)
+savefig('contour_plot_heat_rate.png')
```
Comment (luminoctum) [inline python_codes/mass_budget.py]: Right now, the program does not output \dot q_c. You can comment out this paragraph. We'll do the q_c part later

File: python_codes/mass_budget.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+
+
+show()
```
Comment (luminoctum) [inline python_codes/mass_budget.py:76]: How do you run this program?
```
./mass_budget.py
```
?

File: python_codes/contour_plots.py [hunk 10]
```diff
@@ -43,13 +43,18 @@
 # NH3 Variations
 
 NH3avg = mean(NH3_total, axis = (0,2))
-print(NH3_total.shape, NH3avg.shape)
 NH3_variations = NH3_total - NH3avg.reshape((1,nx1,1))
```
Comment (luminoctum) [inline python_codes/contour_plots.py]: Let's formalize the naming convention. 
Time: **time**
Temperature: **temp**
Pressure: **pres**
NH3 vapor mass mixing ratio: **qNH3v**
NH3 vapor density: **rhoNH3v**
NH3 cloud density: **rhoNH3c**
NH3 cloud mass mixing ratio: **qNH3c**
NH3 precipitation mass mixing ratio: **qNH3p**
H2O total mass mixing ratio: **qH2Ot**
temporal average: **_bar**
horizontal average: **_avg**
anomaly with respect to some reference value: **a**
so **NH3_variations** should be changed to **qNH3a**

File: python_codes/contour_plots.py [hunk 11]
```diff
@@ -43,13 +43,18 @@
...
 
+NH3_0 = NH3_total[0,:,0]
+NH3_variations_from_initial = NH3_total - NH3_0.reshape((1,nx1,1))
```
Comment (luminoctum) [inline python_codes/contour_plots.py]: Now you have two anomalies, one from the average, another from the initial condition. Let's name them in the following way:
anomaly from average: **qNH3a1**
anomaly from the initial condition: **qNH3a2**

File: python_codes/contour_plots.py [hunk 12]
```diff
@@ -43,13 +43,18 @@
...
+
+
 #H2O Variations
 
-H2Oavg = mean(H2O_total, axis = (0,2))
+H2O_0 = H2O_total[0,:,0]
+H2O_variations_from_initial = H2O_total - H2O_0.reshape((1,nx1,1))
 
+H2Oavg = mean(H2O_total, axis = (0,2))
 H2O_variations = H2O_total - H2Oavg.reshape((1,nx1,1))
```
Comment (luminoctum) [inline python_codes/contour_plots.py]: Do the same naming for H2O

File: python_codes/contour_plots.py [hunk 13]
```diff
@@ -59,37 +64,36 @@
 
 fig,ax=subplots(1,1)
 
-contour_plot_temp_var = ax.contourf(Time, Pressure, sum(temp_variations, axis =2))
+contour_plot_temp_var = ax.contourf(Time, Pressure, mean(temp_variations, axis =2))
```
Comment (luminoctum) [inline python_codes/contour_plots.py]: rewrite that to:
```
h = ax.contourf(Time, Pressure, mean(temp_variations, axis =2))
```

File: python_codes/contour_plots.py [hunk 14]
```diff
@@ -59,37 +64,36 @@
...
 ax.set_yscale('log')
 fig.colorbar(contour_plot_temp_var)
-savefig('contour_plot_temp.png')
+savefig('contour_plot_temp_deep.png')
 
 fig,ax=subplots(1,1)
 
-contour_plot_H2O = ax.contourf(Time, Pressure, sum(H2O_variations, axis = 2))
+contour_plot_H2O = ax.contourf(Time, Pressure, sum(H2O_variations_from_initial, axis = 2))
 title(r'$H_2O$ Concentration')
```
Comment (luminoctum) [inline python_codes/contour_plots.py]:
```
ax.set_title
```

File: python_codes/contour_plots.py [hunk 15]
```diff
@@ -59,37 +64,36 @@
...
 ax.set_xlabel(r'Time(s)',fontsize = 12)
 ax.set_ylabel(r'Pressure',fontsize = 12)
-ax.set_ylim(max(pavg), min(pavg))
+ax.set_ylim(max(pavg), 10**6)
 ax.set_yscale('log')
 fig.colorbar(contour_plot_H2O)
```
Comment (luminoctum) [inline python_codes/contour_plots.py]:
```
fig.colorbar(h)
```

Comment (luminoctum) [reply python_codes/contour_plots.py]: If you need multiple plots, use **h1**, **h2**

File: python_codes/contour_plots.py [hunk 16]
```diff
@@ -59,37 +64,36 @@
...
-savefig('contour_plot_H2O.png')
+savefig('contour_plot_H2O_deep.png')
```
Comment (luminoctum) [inline python_codes/contour_plots.py]:
```
savefig('contour_plot_H2O_deep.png', bbox_inches = 'tight')
```

File: python_codes/mass_budget.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+ax.set_yscale('log')
+fig.colorbar(contour_plot_mass_flux)
+savefig('contour_plot_mass_flux.png')
+
+
+#Plotting the Heating Rate due to condensation? 
+# unsure what the term is called without the Latent heat
+
+fig,ax= subplots(1,1)
+contour_plot_heat_rate = ax.contourf(Time, Pressure, heat_rate)
```
Comment (luminoctum) [inline python_codes/mass_budget.py]:
```
h = xxx
```

File: python_codes/mass_budget.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+title(r'$-<\rho \dot{q_c}>$')
```
Comment (luminoctum) [inline python_codes/mass_budget.py]: set_title

File: python_codes/contour_plots.py [hunk 19]
```diff
@@ -59,37 +64,36 @@
...
 
-contour_plot_temp_var = ax.contourf(Time, Pressure, sum(temp_variations, axis =2))
+contour_plot_temp_var = ax.contourf(Time, Pressure, mean(temp_variations, axis =2))
 title(r'Temperature Variations')
 ax.set_xlabel(r'Time(s)',fontsize = 12)
 ax.set_ylabel(r'Pressure',fontsize = 12)
-ax.set_ylim(max(pavg), min(pavg))
+ax.set_ylim(max(pavg), 10**6)
 ax.set_yscale('log')
 fig.colorbar(contour_plot_temp_var)
```
Comment (luminoctum) [inline python_codes/contour_plots.py]: fig.colorbar(h)

File: python_codes/contour_plots.py [hunk 20]
```diff
@@ -59,37 +64,36 @@
...
-savefig('contour_plot_temp.png')
+savefig('contour_plot_temp_deep.png')
```
Comment (luminoctum) [inline python_codes/contour_plots.py]:
```
savefig('xxx', bbox_inches = 'tight')
```

File: python_codes/contour_plots.py [hunk 21]
```diff
@@ -59,37 +64,36 @@
...
 
 fig,ax=subplots(1,1)
 
-contour_plot_H2O = ax.contourf(Time, Pressure, sum(H2O_variations, axis = 2))
+contour_plot_H2O = ax.contourf(Time, Pressure, sum(H2O_variations_from_initial, axis = 2))
```
Comment (luminoctum) [inline python_codes/contour_plots.py]: Normally, we use lower case words for a variable. so change Time to **time**, Pressure to **pres**

**Review by luminoctum — CHANGES_REQUESTED** [review]
Let me know how your run these two programs. I'll set up continuous integration.

**Comment by mayaafields** [comment]
contour_plot_density

**Review by luminoctum — APPROVED** [review]

**Comment by luminoctum** [comment]
I'll merge the request. We can discuss the figure later.

**Pull request merged**